In [1]:
import os    
from dotenv  import load_dotenv
from pathlib import Path
import requests
import pandas as pd
pd.options.display.max_colwidth = 600  # Set Full Display

'ASHWOOD', 'BALWYN', 'BALWYN NORTH', 'BLACKBURN',
'BLACKBURN SOUTH', 'BOX HILL', 'BOX HILL SOUTH', 'BURWOOD',
'BURWOOD EAST', 'CAMBERWELL', 'CHADSTONE', 'CLAYTON',
'CLAYTON SOUTH', 'DONCASTER', 'DONCASTER EAST', 'FOREST HILL',
'GLEN IRIS', 'GLEN WAVERLEY', 'HUNTINGDALE', 'MITCHAM',
'MONT ALBERT', 'MOUNT WAVERLEY', 'MULGRAVE', 'NOTTING HILL',
'NUNAWADING', 'OAKLEIGH', 'OAKLEIGH EAST', 'RINGWOOD',
'SURREY HILLS', 'TEMPLESTOWE', 'VERMONT', 'VERMONT SOUTH',
'WHEELERS HILL'

In [2]:
suburb="Glen_Waverley"

## First Key:

In [7]:
load_dotenv(dotenv_path = Path("C:/Users/Leon/API_keys/.env"))
auproperty_api=os.getenv('auproperty_vincent2')
type(auproperty_api)

str

In [8]:
auth_url = 'https://api2.aupropertyreport.com/api/auth'

auth_response = requests.post(auth_url, data = {
                        'email':"test@gmail.com",
                        'api_key':auproperty_api,
                        })
auth_token = auth_response.json() 
access_token = auth_token['data']["token"]
auth = {"Authorization":"Bearer " + access_token}
auth

{'Authorization': 'Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL2FwaTIuYXVwcm9wZXJ0eXJlcG9ydC5jb20vYXBpL2F1dGgiLCJpYXQiOjE2NTEyNTAyMDksImV4cCI6MTY1MTMzNjYwOSwibmJmIjoxNjUxMjUwMjA5LCJqdGkiOiJzc0dGVm5QUEFzUmFURlF3Iiwic3ViIjozNywicHJ2IjoiNDBhNDZkZjhmNTlkN2E5ZGVkZDc0NjRkZDI2NTNjOGIxNzE2ZTBiZCIsInBsYW4iOm51bGx9.h-088z23X4vO-hxcwU6ZBJrdEhJ8q38ZQX_jhIe7LgE'}

In [10]:
property_search_url = "https://api2.aupropertyreport.com/api/property/search?q=3+17+hillcrest+ave+chadstone+vic+3148"

# GET Request
property_search_request = requests.get(property_search_url,headers=auth,params={"limit":"50"})
property_search_respond = property_search_request.json()
property_search_respond

{'message': 'You have exceeded your rate limit.', 'status_code': 429}

Exceed: {'message': 'You have exceeded your rate limit.', 'status_code': 429}

Normal:  [{'address': '3/17 Hillcrest Avenue, Chadstone Vic 3148',
  'code': 'unit-3-17-hillcrest-ave-chadstone-vic-3148',
  'state': 'VIC',
  'suburb': 'Chadstone',
  'postcode': '3148'}]

In [11]:
# 由于 address code search 的 respond 都是 list, 即使是空的，或者多个的，用来check rate-limit
if isinstance(property_search_respond, dict):

    auproperty_api=os.getenv('auproperty_vincent2')

    auth_url = 'https://api2.aupropertyreport.com/api/auth'

    auth_response = requests.post(auth_url, data = {
                        'email':"leon20130301@hotmail.com",
                        'api_key':os.getenv('auproperty_vincent'),
                        })
    auth_token = auth_response.json() 
    access_token = auth_token['data']["token"]
    auth = {"Authorization":"Bearer " + access_token}
    print(auth) # 如果没有任何显示，则说明没有使用这个 API key

{'Authorization': 'Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL2FwaTIuYXVwcm9wZXJ0eXJlcG9ydC5jb20vYXBpL2F1dGgiLCJpYXQiOjE2NTEyNTAyMTUsImV4cCI6MTY1MTMzNjYxNSwibmJmIjoxNjUxMjUwMjE1LCJqdGkiOiJlcjBMajlzZTlLc1pGaWxlIiwic3ViIjozNiwicHJ2IjoiNDBhNDZkZjhmNTlkN2E5ZGVkZDc0NjRkZDI2NTNjOGIxNzE2ZTBiZCIsInBsYW4iOm51bGx9.Zg6gd3pGERSL8VlNwOZ8zwN8TMTALoLYTUlAxAhe9ac'}


In [21]:
# 由于 address code search 的 respond 都是 list, 即使是空的，或者多个的，用来check rate-limit
if isinstance(property_search_respond, dict):

    auproperty_api=os.getenv('auproperty_vincent2')

    auth_url = 'https://api2.aupropertyreport.com/api/auth'

    auth_response = requests.post(auth_url, data = {
                        'email':"pengmel@hotmail.com",
                        'api_key':os.getenv('auproperty_vincent3'),
                        })
    auth_token = auth_response.json() 
    access_token = auth_token['data']["token"]
    auth = {"Authorization":"Bearer " + access_token}
    print(auth) # 如果没有任何显示，则说明没有使用这个 API key

{'Authorization': 'Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL2FwaTIuYXVwcm9wZXJ0eXJlcG9ydC5jb20vYXBpL2F1dGgiLCJpYXQiOjE2NTEyNTEyMjEsImV4cCI6MTY1MTMzNzYyMSwibmJmIjoxNjUxMjUxMjIxLCJqdGkiOiJrajJHakFtQkVhb0FOd0NzIiwic3ViIjozOSwicHJ2IjoiNDBhNDZkZjhmNTlkN2E5ZGVkZDc0NjRkZDI2NTNjOGIxNzE2ZTBiZCIsInBsYW4iOm51bGx9.COhZfCeb3GdOpZvZm_n6wIg8hTYgtumm5nPVZUTK5kg'}


# Import Combined Addresscode:

### Initialization: Loading saved data

In [22]:
# use r -- to avoid unicode error !
df_target_path=Path(rf'C:\Users\LEON\test\domain\Address_Code_Output_Combined_{suburb}.xlsx')
df_target = pd.read_excel(df_target_path, parse_dates=True)
print(df_target.info())
df_target.tail(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19479 entries, 0 to 19478
Data columns (total 2 columns):
address         19479 non-null object
address_code    19479 non-null object
dtypes: object(2)
memory usage: 304.5+ KB
None


address  \
19478  11+MANNERING+DRIVE+GLEN+WAVERLEY+3150   

                                 address_code  
19478  11-mannering-dr-glen-waverley-vic-3150

In [23]:
# use r -- to avoid unicode error !
filesaved_path=Path(rf'C:\Users\LEON\test\domain\History_Output_{suburb}.xlsx')
filesaved_df = pd.read_excel(filesaved_path, parse_dates=True)
print(filesaved_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3164 entries, 0 to 3163
Data columns (total 5 columns):
address_check     3154 non-null object
full_history      3154 non-null object
history_only      3154 non-null object
recent_history    2848 non-null object
address_name      3154 non-null object
dtypes: object(5)
memory usage: 123.7+ KB
None


In [24]:
missing_history = filesaved_df["full_history"].str.contains("{").sum() - filesaved_df["recent_history"].str.contains("{").sum()
print(f"There are {missing_history} missing history data so far!")

There are 443 missing history data so far!


# Searching for History:

### Loop over dataframe and make API call:

In [25]:
# df_target.index
# df_target.iloc[0]["address_code"]

In [26]:

for i in range(10): # Run X Times 
    # use r -- to avoid unicode error !
    filesaved_path=Path(rf'C:\Users\LEON\test\domain\History_Output_{suburb}.xlsx')
    filesaved_df = pd.read_excel(filesaved_path, parse_dates=True)

    # Initialization:
    list_address_code=[] # to check if matching
    list_address=[]
    list_full_history=[]
    list_history_only=[]
    list_recent_history=[]
    api_count=0

    # Loop over dataframe and make API call:
    for index in df_target.index:
        each = df_target.iloc[index]["address_code"]

        try:

            if "[" not in each and index >= len(filesaved_df) and api_count <= 50-1: # 非常注意：从0开始，所以call -1个
                property_history_url = f"https://api2.aupropertyreport.com/api/property/{each}/history"
                # GET Request
                property_history_request = requests.get(property_history_url,headers=auth,params={"limit":"50"})
                property_history_respond=property_history_request.json()
                api_count += 1

                if "message" not in property_history_respond:

                    if "[" not in each:
                        list_address_code.append([property_history_respond["data"]["address_code"]])
                        list_address.append([property_history_respond["data"]["address"]])
                        list_full_history.append([property_history_respond])
                        list_history_only.append([property_history_respond["data"]["history"]["data"]])
                
                        if  len(property_history_respond["data"]["history"]["data"])!= 0:
                            list_recent_history.append([property_history_respond["data"]["history"]["data"][0]])
                        else:
                            list_recent_history.append([[]]) #非常注意：因爲沒有像上面if中多套一層【】所以是 null值，空的=是想要的狀態！
                    else:
                        list_address_code.append([[]])
                        list_address.append([[]])
                        list_full_history.append([[]])
                        list_history_only.append([[]])
                        list_recent_history.append([[]])
                else:   
                    print("ERROR IN RESPOND")
                    pass #break
            
            elif "[" in each and index >= len(filesaved_df) and api_count <= 5-1:
                list_address_code.append([[]])
                list_address.append([[]])
                list_full_history.append([[]])
                list_history_only.append([[]])
                list_recent_history.append([[]])

        except:
            print("ERROR in property_history_request!!")
            pass    
                
    # At the end, convert the list to dataframe:
    df_list_address_code = pd.DataFrame (list_address_code, columns = ['address_check'])  
    df_list_address = pd.DataFrame (list_address, columns = ['address_name'])   
    df_list_full_history = pd.DataFrame (list_full_history, columns = ['full_history'])    
    df_list_history_only = pd.DataFrame (list_history_only, columns = ['history_only'])   
    df_list_recent_history = pd.DataFrame (list_recent_history, columns = ['recent_history'])  
    df_temp=pd.concat([df_list_address_code,df_list_address,df_list_full_history,df_list_history_only,df_list_recent_history],sort=False,axis=1)

    # update the master dataframe:
    updated_df=pd.concat([filesaved_df,df_temp],sort=False)
    updated_df_resetindex = updated_df.reset_index(drop=True)

    # create excel writer object
    writer=pd.ExcelWriter(f'History_Output_{suburb}.xlsx')
    # write dataframe to excel
    updated_df_resetindex.to_excel(writer, index=False)
    # save the excel
    writer.save()
    print('DataFrame is written successfully to Excel File.')

    print("-------------------------------------------")
    print(updated_df_resetindex.info())
    print("-------------------------------------------")
    print(f"we have {len(list_history_only)} recorded") # 等于 API call 次数，因为空[]也放进去了
    print(f"the API Used in this run is: {api_count}")
    print(f"lenght of master df is now: {len(updated_df_resetindex)}")


DataFrame is written successfully to Excel File.
-------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
address_check     3204 non-null object
full_history      3204 non-null object
history_only      3204 non-null object
recent_history    2898 non-null object
address_name      3204 non-null object
dtypes: object(5)
memory usage: 125.7+ KB
None
-------------------------------------------
we have 50 recorded
the API Used in this run is: 50
lenght of master df is now: 3214
DataFrame is written successfully to Excel File.
-------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3264 entries, 0 to 3263
Data columns (total 5 columns):
address_check     3254 non-null object
full_history      3254 non-null object
history_only      3254 non-null object
recent_history    2948 non-null object
address_name      3254 non-null object
dtypes: object(5)
memory usage: 127.

PS: if recent history is less than full history, then it means N/A !!!

# Concat address_code and address_history:

In [27]:
# Combine combined_address_code and history
df_combined_history = pd.concat([df_target.iloc[0:len(updated_df_resetindex)],updated_df_resetindex.iloc[0:len(updated_df_resetindex)]],sort=False,axis=1)
display(df_combined_history.tail(1))

# address_check 于 address_code 一致，則 match !!!

address  \
3662  924+HIGH+STREET+ROAD+GLEN+WAVERLEY+3150   

                                   address_code  \
3662  924-high-street-rd-glen-waverley-vic-3150   

                                  address_check  \
3662  924-high-street-rd-glen-waverley-vic-3150   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 full_history  \
3662  {'data': {'state': 'VIC', 'address_code': '924-high-street-rd-glen-waverley-vic-3150', 'gnaf_pid': 'GAVIC420399703', 'address': '924 High Street Road', 'suburb_code': 'glen-waverley-vic-3150', 'suburb_name': 'Glen Waverley', 'postcode': '3150', 'sa2_code': 212051321, 'lat': -37.876399, 'lng': 145.18236, 'image': 'https://i7.auhouseprices.com/img/QTdwTnorM3VoQjZaUE9KRkcwZWpEYzBTZW5ERnRSbzlyRHlGZXRUTlpCd3JkRGFGWEVGa0lpaS82dnNILy9taStnN3NzSWUxYitWNitWbjlDOXlQTTlDc3FCUzV1MFpid3BvYWpCeFpQY1pjVXNwUFZVRzF2ZHN2THNiTU1SakNlNzNqMlFHNGdnOW44TzRrQ0JNQTV2OVJhV2FyTHhOalZJaVVydFZmTG9BPQ/preview.jpg', 'og...   

                                                                                                                                                                                                              history_only  \
3662  [{'event': 'Sold', 'date': '2009-07-25', 'price': 795000, 'agency': 'N R Reid & Co - Waverley'}, {'event': 'Sold', 'date': '2006-09-09', 'price': 340500}, {'event': 'Sold', 'date': '1997-01-20', 'price': 140000}]   

                                                                                      recent_history  \
3662  {'event': 'Sold', 'date': '2009-07-25', 'price': 795000, 'agency': 'N R Reid & Co - Waverley'}   

              address_name  
3662  924 High Street Road

In [28]:
# create excel writer object
writer=pd.ExcelWriter(f'Address_Code_History_Combined_{suburb}.xlsx')
# write dataframe to excel
df_combined_history.to_excel(writer, index=False)
# save the excel
writer.save()
print('DataFrame is written successfully to Excel File.')

DataFrame is written successfully to Excel File.


# Search Check for Null:

In [19]:
'''
property_history_url = "https://api2.aupropertyreport.com/api/property/42-campbell-st-glen-waverley-vic-3150/history"

# GET Request
property_history_request = requests.get(property_history_url,headers=auth,params={"limit":"50"})
property_history_respond=property_history_request.json()
property_history_respond
'''

'\nproperty_history_url = "https://api2.aupropertyreport.com/api/property/42-campbell-st-glen-waverley-vic-3150/history"\n\n# GET Request\nproperty_history_request = requests.get(property_history_url,headers=auth,params={"limit":"50"})\nproperty_history_respond=property_history_request.json()\nproperty_history_respond\n'